unit-tests/dds/test-metadata-syncer.py

```diff
@@ -0,0 +1,330 @@
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 1 )
+    test.check( last_metadata ) and test.check_equal( md_id( last_metadata ), 1 )
+    test.check_equal( len(dropped_metadata), 0 )
+
+with test.closure( 'Enqueue 1 image after some metadata -> drops + match' ):
+    syncer = new_syncer()
+    for i in range(4):
+        syncer.enqueue_metadata( i, new_metadata( i ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_frame( 2, new_image( 2 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 2 )
+    test.check( last_metadata ) and test.check_equal( md_id( last_metadata ), 2 )
+    test.check_equal( len(dropped_metadata), 2 )  # 0 and 1
+
+with test.closure( 'Enqueue 1 image then earlier metadata -> nothing out; metadata is dropped' ):
+    syncer = new_syncer()
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_metadata( 0, new_metadata( 0 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 1 )  # not going to be any frame for it
+
+with test.closure( 'Enqueue 1 image then later metadata -> image out' ):
+    syncer = new_syncer()
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_metadata( 2, new_metadata( 2 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 1 )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+
+with test.closure( 'Enqueue 2 images then matching metadata for the second -> both out' ):
+    syncer = new_syncer()
+    syncer.enqueue_frame( 1, new_image( 1 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_frame( 2, new_image( 2 ) )
+    test.check_equal( last_image, None )
+    test.check_equal( last_metadata, None )
+    test.check_equal( len(dropped_metadata), 0 )
+    syncer.enqueue_metadata( 2, new_metadata( 2 ) )
+    test.check( last_image ) and test.check_equal( image_id( last_image ), 2 )
+    test.check( last_metadata ) and test.check_equal( md_id( last_metadata ), 2 )
+    test.check_equal( len(dropped_metadata), 0 )
+
+with test.closure( 'Enqueue during callback' ):
+    def enqueue_during_callback( image, metadata ):
+        on_frame_ready( image, metadata )
+        id = image_id( image ) + 1
+        syncer.enqueue_frame( id, new_image( id ))
+        # Can generate a 'RuntimeError: device or resource busy' and then the 2nd frame won't come out
```

Comment:
Is this comment still relevant? What does it mean seeing this error, should we just try again?

Reply:
Not relevant any more -- will remove or clarify it's what may happen if it's not working right...

Reply:
Now:
```
# Above used to generate a 'RuntimeError: device or resource busy' so the 2nd frame didn't
# come out - that's what we're testing: the syncer shouldn't be locked inside the callback!
```